config/m4/ucs.m4

```diff
@@ -201,5 +201,20 @@ case ${host} in
     AC_DEFINE([HAVE_HW_TIMER], [1], [high-resolution hardware timer disabled])
 esac
 
+#
+# Enable built-in memcpy
+#
+AC_ARG_ENABLE([builtin-memcpy],
+	AS_HELP_STRING([--enable-builtin-memcpy],
+	               [Enable builtin memcpy routine, default: YES]),
```

Comment:
I'm not convinced that this has to be enabled by default. Typically if you use most recent libc memcpy performance is nearly optimal. In cases when libc is outdated, we should use the builtin one. I'm wandering if this is something that we can detect in runtime and overwrite memcpy on the fly ? @yosefe

Reply:
we saw that on x86 centos7, glibc memcpy is worse for shared memory (because of RFO)
runtime flag is another branch, do you think it's worth it?

Reply:
@yosefe I was think overwriting the address of function, similar to malloc hooks. What is RFO ?

Reply:
Centos7 glibs is old. Eventually, they will update the libc and libc keep improving their memcpy. DPDK also has internal memcpy but it is disabled by default for the above reasons,

Reply:
RFO=Request For Ownership (cache miss on write)
i'd prefer not overwrite memcpy, it probably can't be done in thread safe way and will affect code which is unrelated to ucx (more than munmap hook)
we can check glibc version at configure time to make the default decision

Reply:
@yosefe - maybe we can run simple benchmark during configure and decide ? I don't think version check is reliable.

Reply:
```It is key to note this applies only for the various stores within the single instruction itself: once complete, the block of stores still appear ordered with respect to prior and subsequent stores. So code can see stores from the rep movs out of order with respect to each other but not with respect to prior or subsequent stores (and it's the latter guarantee you usually need). It will only be a problem if you use the end of the copy destination as a synchronization flag, instead of a separate store.```

You have to be really careful how you use ```REP```.

Reply:
maybe if someone is polling on 32-bit flag, it could read invalid value

Reply:
yes, maybe we shell call this relaxed_memcpy or something like this to avoid confusion.

Reply:
More on this:
"REP MOVS instructions are very slow on small blocks because the startup cost is about 35 cycles. If you do plain simple MOV EAX in a loop, there are no startup costs and you can copy lots of data during these 35 cycles."

Reply:
@shamisp Intel compiler generates ```rep movsb``` to copy data & I believe they know how to copy memory fastest. we are planning to use this optimized memcpy for bcopy API only where data length is higher than "max_short"

Reply:
@hoopoepg On conceptual level it is "good thing". I think implementation/naming is a bit misleading since it targets a very narrow use case. Probably should be called bcopy_relaxed_pack() with debug mode check for the size.